src/Metadata/MetadataFactory.php

```diff
@@ -0,0 +1,200 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+declare(strict_types=1);
+
+namespace Doctrine\Annotations\Metadata;
+
+use ReflectionClass;
+use ReflectionProperty;
+
+
+use Doctrine\Annotations\Parser\MetadataParser;
+use Doctrine\Annotations\Annotation\Target;
+
+/**
+ * Annotation metadata factory
+ *
+ * @author Fabio B. Silva <fabio.bat.silva@gmail.com>
+ */
+class MetadataFactory
+{
+    /**
+     * @var MetadataParser
+     */
+    private $parser;
+
+    /**
+     * @var array
+     */
+    private $cache = [];
+
+    /**
+     * Constructor
+     *
+     * @param MetadataParser $parser
+     */
+    public function __construct(MetadataParser $parser)
+    {
+        $this->parser = $parser;
+    }
+
+    /**
+     * Gets the class metadata descriptor for a annotation class.
+     *
+     * @param string $className The name of the class.
+     *
+     * @return ClassMetadata
+     *
+     * @throws \Exception
+     */
+    public function getMetadataFor(string $className)
+    {
+        if (isset($this->cache[$className])) {
+            return $this->cache[$className];
+        }
+
+        $class       = new \ReflectionClass($className);
+        $constructor = $class->getConstructor();
+        $docComment  = $class->getDocComment();
+        $metadata    = new ClassMetadata();
+
+        $metadata->hasConstructor = false;
+        $metadata->class          = $className;
+        $metadata->target         = Target::TARGET_ALL;
+
+        if ($constructor !== null && $constructor->getNumberOfParameters() > 0) {
+            $metadata->hasConstructor = true;
+        }
+
+        $annotations = $this->parser->parseAnnotationClass($class);
+        $indexed     = $this->getIndexedAnnotations($annotations);
+
+        if ( ! $this->isAnnotation($class, $indexed)) {
+            return null;
+        }
+
+        if (isset($indexed['Target'])) {
+            $metadata->target = $indexed['Target']->target;
+        }
+
+        $this->collectPropertiesMetadata($class, $metadata);
+
+        return $this->cache[$className] = $metadata;
+    }
+
+    /**
+     * @param \ReflectionClass $class
+     * @param array            $annotations
+     *
+     * @return bool
+     */
+    private function isAnnotation(ReflectionClass $class, array $annotations) : bool
+    {
+        if ($class->isSubclassOf('Doctrine\Annotations\Annotation')) {
```

Comment:
@lcobucci see my previous comment. Review will be much easier if the switch to ``::class`` is done in a separate PR